tracy-client/src/lib.rs

```diff
@@ -1,32 +1,203 @@
+/// already has been enabled.
+///
+/// # Example
+///
+/// ```rust
+/// fn main() {
+///     tracy_client::enable();
+///     // ...
+///     tracy_client::disable();
+/// }
+pub fn enable() {
+    #[cfg(feature = "enable")]
+    {
+        // We don't particularly care about efficiency here, so we’ll be using CAS loops and
+        // orderings that aren't necessarily the fastest possible in this instance.
+        let old_state = loop {
+            let result = ENABLE_STATE.fetch_update(Ordering::SeqCst, Ordering::SeqCst, |state| {
+                // Here we want to increment the reference count, and also apply the tansition from
+                // state 0 (disabled) to state 1 (enabling), if possible.
+                match state & STATE_MASK {
+                    STATE_DISABLED => Some(state + STATE_ENABLING + 1),
+                    STATE_ENABLED => Some(state + 1),
+                    STATE_ENABLING => Some(state + 1),
+                    // Wait for the ongoing disable to complete.
+                    STATE_DISABLING => None,
+                    _ => unreachable!(),
+                }
+            });
+            if let Ok(result) = result {
+                break result;
+            }
+        };
+        match old_state & STATE_MASK {
+            STATE_DISABLED => {
+                unsafe {
+                    // SAFETY: no known soundness invariants.
+                    sys::___tracy_startup_profiler();
+                }
+                ENABLE_STATE.fetch_add(STATE_STEP, Ordering::SeqCst);
+            }
+            STATE_ENABLING => {
+                // wait for the profiler to reach the enabled state
+                while ENABLE_STATE.load(Ordering::SeqCst) & STATE_MASK != STATE_ENABLED {}
+            }
+            STATE_ENABLED => return,
+            _ => unreachable!(),
+        }
+    }
+}
+
+/// Disable the client.
+///
+/// This function will decrease the activation count and unload the client once this count reaches
+/// zero. Once the client has been disabled, no other calls to the instrumentation APIs may be
+/// made. Note that unloading the client will also discard any data collected up to that point.
+///
+/// When using threads, especially detached ones, consider either never calling `disable`, or at
+/// least use a `enable`-`disable` pair for each thread.
+///
+/// # Examples
+///
+/// ```rust
+/// fn main() {
+///     tracy_client::enable();
+///     // ...
+///     tracy_client::disable();
+/// }
+/// ```
+///
+/// One thing to watch out in particular is drop guards. For example the following code is invalid
+/// because a span guard is dropped after `disable` is called.
+///
+/// ```rust,no_run
+/// {
+///     tracy_client::enable();
+///     let _span = tracy_client::span!("some span");
+///     tracy_client::disable();
+/// } // `_span` is dropped here, after the client has been disabled!
+/// ```
+pub fn disable() {
```

Comment:
From above discussion, it sounds like this must be `unsafe`. Unless calls gracefully panic when disabled?

Reply:
Right, they don’t – in practice the upstream project does not commit to any particular behaviour when the other APIs are called while the client is disabled (which is why we'd observe crashes due to things like UAF before as well.)

That I feel kind of forces our hand to heavily refactor the API of this crate, now that I think about it. So that it isn't possible to call any of the instrumentation before enabling, which I feel kind of leads us back to setting up some context type that would indicate the client is enabled and expose all the other things as methods.

Reply:
A global atomic bool which bypasses the calls when disabled would be another option which better preserves the current pattern, and I think is justifiable in this context as there's global state regardless.